### ISSUE TITLE
Linked library works on default target, but not with `wasm32-unknown-emscripten`

### ISSUE THREAD

**Issue by Guusggg** [emscripten-core/emscripten#17818]
**Please note: Emscripten does compile my Rust code without the external library.**
Edit: I uploaded the complete repo, so it's a little clearer what I'm trying to do.
https://github.com/Guusggg/example-game

Please include the following in your bug report:

**Version of emscripten/emsdk:**
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.20 (5d878c99921ec247d34fb26a20b5a13d60d69e93)
clang version 16.0.0 (https://github.com/llvm/llvm-project 75767a0f9a926641edbef08e31ec2148ff45da67)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: \emsdk\upstream\bin

**Failing command line in full:**
It's a very, very long command with a lot of personal information included so I'm just going to strip it a bit:
```
\\example-game\\target\\wasm32-unknown-emscripten\\debug\\deps\\example_game.js" "-O0" "-g" "-sABORTING_MALLOC=0" "-Wl,--fatal-warnings"
  = note: wasm-ld: error: example-game\target\wasm32-unknown-emscripten\debug\deps\libclient_sys-c3851003994388e0.rlib(my_game.o): undefined symbol: Z_envZ_console_log
          wasm-ld: error: example-game\target\wasm32-unknown-emscripten\debug\deps\libclient_sys-c3851003994388e0.rlib(my_game.o): undefined symbol: Z_envZ_clear_screen
          wasm-ld: error: example-game\target\wasm32-unknown-emscripten\debug\deps\libclient_sys-c3851003994388e0.rlib(my_game.o): undefined symbol: Z_envZ_glActiveTexture
          wasm-ld: error: example-game\target\wasm32-unknown-emscripten\debug\deps\libclient_sys-c3851003994388e0.rlib(my_game.o): undefined symbol: Z_envZ_draw_rect
```

**The issue**
I'm currently trying to compile a WASM2C generated module to use it in Rust. I know Emscripten is not written in nor for Rust, but the toolchain is usually very useable with Rust. 

When I compile it for my native machine, like so:
`cargo run`, the application runs and it loads the WASM2C module perfectly with my handwritten bindings. 
```rust
#[link(name = "client")]
extern "C" {
    #[link_name = "Z_my_gameZ_game_init"]
    pub static game_init: Option<unsafe extern "C" fn() -> u32>;
    #[link_name = "Z_my_gameZ_game_loop"]
    pub static game_loop: Option<unsafe extern "C" fn() -> ()>;    
    #[link_name = "Z_my_gameZ_game_get_score"]
    pub static game_get_score: Option<unsafe extern "C" fn() -> u32>;
}

#[no_mangle]
#[used]
#[export_name = "Z_envZ_draw_rect"]
pub static mut draw_rect: Option<fn(u32, u32, u32, u32) -> ()> = Some(|_x, _y, _width, _height| {
    println!("draw_rect called!");
});
```

But when I use the Emscripten toolchain, it fails with the `undefined symbols` error. I've been getting comfortable with compilers lately and usually I know it's my fault. It's just that when I compile it on my machines target, everything works. But when I try to compile it for the `wasm32-unknown-emscripten` target, it does not link properly and cannot find the symbols. 

Sorry for posting such a stupid issue, I really appreciate all the work you guys do with Emscripten. It's an amazing toolchain! I'm completely at a loss of why it runs on my machines target, but doesn't on the wasm32 target...

Thank you for reading this. 

Kind regards,

**Comment by Guusggg** [comment]
After thinking about it a while, I'm not sure if this is the best place to post this as an issue.

**Comment by sbc100** [comment]
It looks like you might be trying to compile the C code generated by wasm2c using emcc, is that right?  Normally one uses the native host compiler to build that code.   So it would go something like  C/C++ source -> emcc -> wasm file -> wasm2c -> .c file -> native toolchain

Perhaps you could explain a little bit more what you are trying to do?

**Comment by Guusggg** [comment]
@sbc100 Hi, thank you for your reaction! I really wanted to use Rust in this project but because I'm to novice in the language so far I've decided to drop the idea. 

You are definitely correct that I'm trying to use wasm2c's generated code in a Rust project. Basically the setup is:
Rust project with `build.rs` that compiles `wasm2c`'s generated code to a static library, which I'm trying to link into the Rust project, targeting `emscripten` to run it (The input to `wasm2c` is using `emscripten` heavily, so instead of emulating it I choose to just build it in emscripten and glue it all together). 

This works absolutely fantastic in a `cpp` project, but I'm unable to resolve the undefined symbols in Rust. I've closed this as I concluded it's my inexperience with Rust. 

I could compile the Rust project for native, and it works very well. However when I try to target `wasm32-unknown-emscripten` (as the original binary relies heavily on emscripten, it makes it easier for me to emulate) I'm getting undefined symbols. 

Basically I'm unable to link Rust's symbols to the static library from `wasm2c` for the `wasm32-unknown-emscripten` target, which I found strange as it did compile for my native target.

Please don't spend too much thoughts on this though, I have a working project in `cpp` but I like Rust a lot so I thought it would be a good projec to tackle in Rust.

**Comment by sbc100** [comment]
Wow, that is an interesting setup.   Where does the input wasm that are you supplying to wasm2c come from?  When you are targeting emscripten it might be simpler just to load the wasm file and run it rather than using wasm2c at all?   I

f that doesn't work and you want to continue down this route then you need to look at why the `#[link_name = "Z_my_gameZ_game_get_score"]` things don't seem to be working under emscirpten.  Perhaps the `link_name` think is relying on some linker feature that isn't working correctly in the wasm toolchain.   Could you attach the object file that results from compiling the file that contains those lines?

**Comment by Guusggg** [comment]
@sbc100 Haha, yes it sure is a "uncommon" but I solemnly swear I'm up to no good ;). 

You can view the project on https://github.com/Guusggg/example-game
I tried to attach the `libclient.a` which is built from `my_game.o` and `wasm-rt-impl.o`, but GitHub won't take it.
Unluckily I'm on Windows for the forseeable time, so I don't really have access to inspect the object files or archive files...

I think there's a linker flag that I need to set for the emscripten linker, although I assumed it was a drop-in replacement...

**Comment by sbc100** [comment]
You can use llvm tools (which come with both emscripten toolchain and I believe the rust toolchain) to inspect object files.  For example, you can use `llvm-nm` to list symbols present.
